zip-0222.rst

```diff
@@ -0,0 +1,322 @@
+>= 1     ``tx_out``             list of outputs             ``vector``
+>= <TBD> ``tze_in_count``       variable-length integer     ``compactSize``
+>= <TBD> ``tze_in``             list of TZE inputs          ``vector``
+>= <TBD> ``tze_out_count``      variable-length integer     ``compactSize``
+>= <TBD> ``tze_out``            list of TZE outputs         ``vector``
+>= 1     ``lock_time``          block height or timestamp   ``uint32``
+... as before
+======== ====================== =========================== ===============
+
+Both ``tze_in`` and ``tze_out`` vectors make use of the common serialized
+form ``tze_data`` described below. Serialization of all integral and vector
+types is as with Bitcoin.
+
+``tze_data`` encoding:
+
+====================== ==================================== =====================
+Field                  Description                          Type           
+====================== ==================================== =====================
+``tze_id``             extension ``type``                   ``compactSize``
+``tze_mode``           extension ``mode``                   ``compactSize``
+``tze_data_len``       length of precondition/witness data  ``compactSize``
+``tze_data``           serialized precondition/witness data ``witness_len`` bytes
+====================== ==================================== =====================
+
+TZE Input Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``prevout_hash``       hash of previous txid                ``uint256``    
+``prevout_in``         index into previous txn's outputs    ``uint32``
+``witness``            witness for prevout's precondition   ``tze_data``
+====================== ==================================== ===============
+
+
+TZE Output Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``amount``             spendable amount, in zatoshi         ``int64``    
+``precondition``       precondition encumbering ``amount``  ``tze_data``
+====================== ==================================== ===============
+
+
+Consensus rules
+---------------
+
+Once this ZIP becomes active, the following new consensus rules are enforced:
+
+- For each ``(outpoint, witness)`` pair in ``tze_inputs``:
+
+  - ``outpoint`` MUST reference a precondition of the same type and mode in an already-mined
+    transaction.
+  - ``tze_verify(mode, precondition, witness, context)`` MUST return ``true``.
+
+- If a transaction has non-empty ``tze_inputs`` and non-empty ``tze_outputs``, then every
+  element in both fields MUST have the same ``type``.
+
```

Comment:
What about only tze_inputs, or only tze_outputs? Can the elements be different types?